jme3-plugins/src/gltf/java/com/jme3/scene/plugins/gltf/TextureTransformExtensionLoader.java

```diff
@@ -0,0 +1,150 @@
+            VertexBuffer tc = mesh.getBuffer(VertexBuffer.Type.TexCoord);
+            if (tc == null) {
+                throw new IllegalStateException("The mesh has no texture coordinates");
+            }
+            if (tc.getFormat() != VertexBuffer.Format.Float) {
+                throw new UnsupportedOperationException("Only float texture coord format is supported");
+            }
+            if (tc.getNumComponents() != 2) {
+                throw new UnsupportedOperationException("Only 2D texture coords are supported");
+            }
+            FloatBuffer fb = (FloatBuffer) tc.getData();
+            fb.clear();
+            for (int i = 0; i < fb.limit() / 2; i++) {
+                float x = fb.get();
+                float y = fb.get();
+                fb.position(fb.position() - 2);
+                Vector3f v = m.mult(new Vector3f(x, y, 1));
+                fb.put(v.getX()).put(v.getY());
+            }
+            fb.clear();
+            tc.updateData(fb);   
+        }
+    }
+    
+    // The algorithm relies on the fact that the GltfLoader.class object 
+    // loads all textures of a given geometry/mesh before doing so 
+    // for the next geometry/mesh.    
+    @Override
+    public Object handleExtension(GltfLoader loader, String parentName, JsonElement parent, JsonElement extension, Object input) throws IOException {
+        if (input instanceof Texture2D) {
+            Geometry geom = loader.fetchFromCache("lastCreatedGeom", 0, Geometry.class);
+            if (geom != null) {
+                Matrix3f t = new Matrix3f();
+                Matrix3f r = new Matrix3f();
+                Matrix3f s = new Matrix3f();
+                JsonObject jsonObject = extension.getAsJsonObject();
+                if (jsonObject.has("offset")) {
+                    JsonArray jsonArray = jsonObject.getAsJsonArray("offset");
+                    t.set(2, 0, jsonArray.get(0).getAsFloat());
+                    t.set(2, 1, jsonArray.get(1).getAsFloat());                    
+                }
+                if (jsonObject.has("rotation")) {
+                    float rad = jsonObject.get("rotation").getAsFloat();
+                    r.set(0, 0, (float) Math.cos(rad));
+                    r.set(0, 1, (float) Math.sin(rad));
+                    r.set(1, 0, (float) -Math.sin(rad));
+                    r.set(1, 1, (float) Math.cos(rad));
+                }                
+                if (jsonObject.has("scale")) {
+                    JsonArray jsonArray = jsonObject.getAsJsonArray("scale");
+                    s.set(0, 0, jsonArray.get(0).getAsFloat());
+                    s.set(1, 1, jsonArray.get(1).getAsFloat());
+                }     
+                if (jsonObject.has("texCoord")) {
+                    logger.log(Level.WARNING, "KHR_texture_transform extension: the texCoord property is not supported");                
+                }                 
+                Matrix3f m = t.mult(r).mult(s);

```

Comment:
Better to rename `m` to `transform` or `transformMatrix`.

Reply:
KHR_texture_transform is usually defined at the level of a texture in a glTF model, as it seems to be the case for Blender. Accordingly, GltfLoader calls the extension loader (readExtensionAndExtras) passing a Texture2D to the "input" parameter when reading a texture (readTexture). In this case, the "if" condition for the Mesh of the proposed code above will always be evaluated as false within the extension loader:

if (input instanceof Mesh) { // => always false for most common glTF models

So the extension loader needs an alternative way to get the mesh. The cache of the GltfLoader can be used for this purpose, but meshes are cached too late (created at line 384, cached at line 496 in geomArray) after textures are already read. Having a more complete state in the cache that extension loaders can use seems reasonable, either by modifying the way GltfLoader caches meshes or by adding an extra line in GltfLoader to cache the mesh before textures are read (as done in the latest commit).

However, I believe the proper way to handle this would be for the extension loader to store each texture transform within the corresponding Texture2D object for later processing, but JME does not seem to support such a design/implementation especially due to the fact that textures are meant to be shared among materials and meshes/geometries. 

Indeed, the original purpose of KHR_texture_transform is to support packing textures into a texture atlas, as explained at:
https://github.com/KhronosGroup/glTF/tree/main/extensions/2.0/Khronos/KHR_texture_transform

Also, different transforms might be specified for different textures. For example, see the UV Mapping section of the glTF documentation of Blender, explaining how Mapping nodes can be attached to Image Texture nodes:
https://docs.blender.org/manual/en/latest/addons/import_export/scene_gltf2.html#id18

Reply:
Thanks for the detailed explanation.

>the "if" condition for the Mesh of the proposed code above will always be evaluated as false

Ah, yes you are right, stupid me!

>Indeed, the original purpose of KHR_texture_transform is to support packing textures into a texture atlas

By the way, JME also has a utility to batch geometries and generate a texture atlas.

https://github.com/jMonkeyEngine/jmonkeyengine/blob/417509eb17722a39a941de66d03e078d542d6e8a/jme3-core/src/tools/java/jme3tools/optimize/GeometryBatchFactory.java#L97